### PR TITLE
Add support for html in chat

### LINKED ISSUE #7: How to insert markdown in the chat message() function

### LINKED ISSUE #16: How to show audio component
Hi,

Im trying to show audio component but nothing works.

I tried:

```
st.session_state.history.append({"message": st.audio(audio_bytes, format='audio/mp3'), "is_user": True, "key":str(uuid.uuid4())})
```

And:

```
st.session_state.history.append({"message": '<audio controls src="audio.mp3"></audio>', "is_user": True, "key":str(uuid.uuid4())})
```

The HTML show as text.

How to make it work?

### PR DESCRIPTION
**Pull request by 0toshigami**
The chat widget currently just support text only, so I decide to add a package to parse HTML string to React Component in safe way.

Thanks for this awesome lib: [react-html-parser](https://github.com/peternewnham/react-html-parser)

#### Demo
image

### CODE REVIEW

**Comment by paulocoutinhox** [comment]
top!

File: streamlit_chat/frontend/src/stChat.tsx [hunk 1]
```diff
@@ -71,7 +72,7 @@ class Chat extends StreamlitComponentBase {
     return (
       <Chat isUser={isUser}>
         <Avatar src={avatarUrl} alt="profile" draggable="false"/>
-        <Message>{message}</Message>
+        <Message>{ReactHtmlParser(message)}</Message>
```
Comment (yashppawar) [inline streamlit_chat/frontend/src/stChat.tsx:75]: Can you add a parameter which will enable/disable this, like the allow_unsafe_html parameter

**Review by yashppawar — CHANGES_REQUESTED** [review]
please add parameter `allow_unsafe_html` or `allow_html` to conditionally use ReactHtmlParser

**Comment by yashppawar** [comment]
@0toshigami is there a way to also provide custom CSS? does ReactHtmlComponent support CSS

also please mention me so I will be notified, for some reason I don't receive any notifications, need to look into that

File: streamlit_chat/frontend/src/stChat.tsx (hunk 1 above)
Comment (0toshigami) [reply streamlit_chat/frontend/src/stChat.tsx]: @yashppawar 
As your suggestion, I've added `allow_html` parameter.

**Comment by 0toshigami** [comment]
> @0toshigami is there a way to also provide custom CSS? does ReactHtmlComponent support CSS
> 
> also please mention me so I will be notified, for some reason I don't receive any notifications, need to look into that

@yashppawar What is meant by custom CSS? Is it for the chat widget or for the content within the chat widget?
Additionally, I have provided support for Markdown syntax.

**Comment by yashppawar** [comment]
> @yashppawar What is meant by custom CSS? Is it for the chat widget or for the content within the chat widget?

sorry for not being clear, I meant CSS for the HTML added inside the chat widget, I saw the video element which had a width greater than that of the chat widget... so was asking if its possible to give CSS to make it look better

**Comment by 0toshigami** [comment]
> sorry for not being clear, I meant CSS for the HTML added inside the chat widget, I saw the video element which had a width greater than that of the chat widget... so was asking if its possible to give CSS to make it look better

@yashppawar Oh, I also noticed that. The video element is an iframe element that has width and height attributes. This makes it convenient for us to customize those attributes. Inline CSS will also work. 

image

**Comment by YiSays** [comment]
Please merge the PR, this is what I need now. 👍

**Comment by yashppawar** [comment]
Sorry was busy due to exams, I was trying to get this to work on my machine, but the HTML was not rendered... is there anything required for this to work? I did delete node modules and install them again

**Comment by 0toshigami** [comment]
> Sorry was busy due to exams, I was trying to get this to work on my machine, but the HTML was not rendered... is there anything required for this to work? I did delete node modules and install them again

Have you read the updated example in my readme? The message method now has 2 new parameters, 
`allow_html` (default = `False`) and 
`is_table` (default = `False`)
```py
st.session_state.setdefault(
    'generated', 
    [{
        'type': 'normal', 'data': f'<img width="100%" height="200" src="{img_path}"/>'
    }]
message(
    message=st.session_state['generated'][0]['data'],
    key=0, 
    allow_html=True,
    is_table=True if st.session_state['generated'][0]['type']=='table' else False
)
```

**Comment by yashppawar** [comment]
yes @0toshigami, I have read it and was trying to execute the same code which you have added in the `readme.md`

**Comment by 0toshigami** [comment]
@yashppawar hmmm, that's strange. Have you tried rebuilding the frontend?

**Comment by brycexu** [comment]
> yes @0toshigami, I have read it and was trying to execute the same code which you have added in the `readme.md`

It works on my machine though.

**Comment by yashppawar** [comment]
Let me try it again today, @0toshigami how do I rebuild frontend? Purge node modules and build dir, and npm i, then npm start?

**Comment by 0toshigami** [comment]
> how do I rebuild frontend? Purge node modules and build dir, and npm i, then npm start?

@yashppawar Maybe you could try to remove the node_modules and package-lock.json, then `npm i`, and `npm run build`

**Comment by duanyongyao** [comment]
I urgently need this feature, please expedite the PR merge. Thank you!

**Comment by stephenleo** [comment]
Looking forward to this PR merge too!

**Comment by 0toshigami** [comment]
> I tried adding an href to an external site and encountering some weird behavior. Is it possible to route the link to open in a new tab pls? @0toshigami

Hi @stephenleo
Because each message is displayed as an `<iframe>`, it's good to use the `target` attribute for `<a>` tag. If you want to open the link in a new tab, use "_blank". Or you can use "_parent" if you one to open the link in the parent window.
`<a href="https://www.google.com" target="_blank">link text</a>`

**Comment by stephenleo** [comment]
amazing! Thank you very much

**Comment by repollo** [comment]
So are we merging this so i can remove the link to requirements.txt instead of just having the library name? Whats are missing to merge?

**Comment by stephenleo** [comment]
I've been using this fork fine for a few days now. Will be a great addition to the library!

**Comment by yashppawar** [comment]
Really sorry for the inconvenience I am causing, I was not able to look at it for long, and might not be able to do more for a week at least (exams), I'll try what I can do..., is it working for everyone else? somehow I end up with more errors than I started with while rebuilding, might have to try a fresh install of node next

**Comment by stephenleo** [comment]
@yashppawar I tried it in a fresh conda environment and it worked great!

**Comment by imsadiku** [comment]
I'm really looking forward that this merge request is accepted!

**Comment by ketank1000** [comment]
Can you please release this to latest version. this change is not available on pypi

**Comment by saurabhgssingh** [comment]
> The chat widget currently just support text only, so I decide to add a package to parse HTML string to React Component in safe way.
> 
> Thanks for this awesome lib: [react-html-parser](https://github.com/peternewnham/react-html-parser)
> 
> #### Demo
> image

Please merge this PR

**Comment by nubgamerz** [comment]
This is also something I desperately need!
Please look at merging the PR at earliest convenience, thanks!

**Comment by damin604** [comment]
Hi @0toshigami ,

Thank you for adding the HTML functionality to the `message` function. While using it, I came across few use-cases where the current front-end output needs improvement.

Using [Collapsed sections](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-collapsed-sections) with markdown,
<details>

<summary>Example code</summary>

```python
import streamlit as st
from streamlit_chat import message

def on_input_change():
    user_input = st.session_state.user_input
    st.session_state.past.append(user_input)
    st.session_state.generated.append("The messages from Bot\nWith new line")

def on_btn_click():
    del st.session_state.past[:]
    del st.session_state.generated[:]

markdown = "Below is an example of collapsible markdown taken from Wikipedia."
paragraph = "Euclidean geometry is a mathematical system attributed to ancient Greek mathematician Euclid, which he described in his textbook on geometry; Elements. Euclid's approach consists in assuming a small set of intuitively appealing axioms (postulates) and deducing many other propositions (theorems) from these. Although many of Euclid's results had been stated earlier,[1] Euclid was the first to organize these propositions into a logical system in which each result is proved from axioms and previously proved theorems."
markdown += """<details><summary>Toggle me!</summary>{paragraph}</details>""".format(paragraph=paragraph)

st.session_state.setdefault('past', ['Show example of collapsible markdown'])
st.session_state.setdefault('generated', [{'type': 'normal', 'data': f'{markdown}'}])

st.title("Example")
with st.container():    
    for i in range(len(st.session_state['generated'])):                
        message(st.session_state['past'][i], is_user=True, key=f"{i}_user")
        message(
            st.session_state['generated'][i]['data'], 
            key=f"{i}", 
            allow_html=True,
            is_table=True if st.session_state['generated'][i]['type']=='table' else False
        )
    
    st.button("Clear message", on_click=on_btn_click)

with st.container():
    st.text_input("User Input:", on_change=on_input_change, key="user_input")
```

</details>

image

Can you please add functionality to support **scrolling** with `message` function?
https://discuss.streamlit.io/t/scrolling-text-containers/26485/3

It will be also useful to have a fixed height container with scrolling for the whole chat functionality. Below is a [Gradio example](https://gradio.app/creating-a-chatbot/),
image

---
@nubgamerz @saurabhgssingh @imsadiku @ketank1000 
Until the PR is merged, you can install the PR using the below command,
```bash
pip install git+https://github.com/0toshigami/st-chat.git#egg=streamlit-chat
```

**Comment by onglette** [comment]
Hello! @0toshigami 

Love the enhancement! Hope PR goes through soon.

My chatbot will cite the source which is a url when responding. and the url is automatically made into a hyperlink after installing the PR. I want the hyperlink to open in a new window but am unable to add target="_blank" to open to a new window. Any suggestions/help would be greatly appreciated!

Thanks

**Comment by eshaanagarwal** [comment]
Hey it would be great if you could merge this PR.

**Comment by michaelfeil** [comment]
Great work. Can you merge this @YashVardhan-AI ?

**Review by yashppawar — APPROVED** [review]

**Comment by yashppawar** [comment]
merged, and published at https://pypi.org/project/streamlit-chat/0.1.0/

**Comment by DonjiKong** [comment]
How can I make hyperlinks work? I just tried it and it did not work

**Comment by yashppawar** [comment]
@DonjiKong I have updated it, now the links should open in new tab

**Comment by jmtatsch** [comment]
I also have the issue that the message bubble does not resize properly with the output when using markdown/html.
Was anyone able to fix that?

**Comment by yashppawar** [comment]
> I also have the issue that the message bubble does not resize properly with the output when using markdown/HTML. Was anyone able to fix that?

hey @jmtatsch, it should resize, to avoid this issue I have tried to call `st.setFrameHeight()` on multiple events

unfortunately, there are many limitations with this component, and streamlit has recently (in v1.24) released `st.chat_message` & `st.chat_input` ( [docs here](https://docs.streamlit.io/library/api-reference#chat-elements) ) which is much more flexible (and fast), you can try that out too if this does not work

**Pull request merged**